### PR TITLE
support cloud-init network-config

### PR DESCRIPTION
**Pull request by michaelkoetter**
This adds basic support for a cloud-init [`network-config` file](https://cloudinit.readthedocs.io/en/18.2/topics/network-config-format-v1.html).

network-config can be used for advanced network configuration, such as VLANs, which can't be configured using the default `dhcpcd`.
The provided default results in the same config as without this change, dhcp for eth0 (plus it won't have any effect because of the bug described below).

To make this work, a few additional steps are needed:

- cloud-init creates the file `/etc/network/interfaces.d/50-cloud-init.cfg`. The intent is that `ifup` etc. should pick this up. Unfortunately they won't, because according to the man page, files in include folder must match the regex `^[a-zA-Z0-9_-]+$.` - so it has to be renamed, or `/etc/network/interfaces` has to be [updated](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=867921)
- By default, `dhcpcd` tries to manage all network interfaces, so there will be a conflict when `ifup` attempts to configure the same interfaces (it will probably do nothing). The workaround is to disable `dhcpcd` to manage all interfaces the old-fashioned way.

Add the following to `user-data` to make this work:

```
runcmd:
  - 'mv /etc/network/interfaces.d/50-cloud-init.cfg /etc/network/interfaces.d/50-cloud-init'
  - 'systemctl disable dhcpcd'

power_state:
  mode: reboot
```

I know this is not very elegant yet. Disabling `dhcpcd` by default would get rid of one of the extra steps above, but then there would still be the bug which results in the "wrong" filename.
Let me know if you have any ideas on how to improve this 😄

### CODE REVIEW

**Comment by michaelkoetter** [comment]
Yes I saw that sample in the flash repo - that seems to be the official way in Raspbian now, at least as long as `dhcpcd` is used.
It seems that most distros are replacing the old `ifupdown` and `/etc/network/interfaces` scripts with new fancy stuff (such as `dhcpcd`, `systemd-networkd` or Ubuntu's `netplan`). While these work fine for a desktop, where 99% of users will just use DHCP, they make it a lot harder to manually configure things IMHO. 

For example I couldn't find a way to configure VLANs with the default `dhcpcd` in Raspbian, while that's easy to accomplish using `/etc/network/interfaces`.
Cloud-Init's network-config will also not work with `dhcpcd` (it could work with `netplan` but I don't think that's supported in Debian). But nevertheless cloud-init will always create the `50-cloud-init.cfg` file unless there is explicit config that tells it not to do that. Right now it just doesn't have any effect because of the source-folder bug.

So if we fix that source-folder bug, I think we also need to include a minimal `network-config` file with the content
```
config: disabled
```
to prevent cloud-init from creating the `50-cloud-init.cfg`.

Then the users could decide whether they want to use network-config (and disable `dhcpcd`) or just use the default (which is probably fine for the majority).

Disabling `dhcpcd` by default would also work, with the network-config I included it will do DHCP on eth0, just that it's now done by `ifup`.

One more issue that I found is that `/etc/resolv.conf` is now managed by `systemd-resolved`, which also doesn't play nice with `ifupdown` - so any resolvers in network-config will not end up there. Still trying to figure out how to fix that...

Basically it's all a big mess... I just want my good old `/etc/network/interfaces` back 😄

**Comment by michaelkoetter** [comment]
I disabled `dhcpcd` and `systemd-resolved` in the chroot-script, installed `resolvconv` (which works with the `ifup` script to manage DNS) and added a fix for `/etc/network/interfaces`.
Now everything works out of the box. The default config will just use DHCP on eth0, so no change from the users perspective.
Custom network-config can be supplied if needed. I tested with the following `network-config` to setup a VLAN interface, static IP and DNS:

```
version: 1
config:
  - type: physical
    name: eth0
  - type: vlan
    name: eth0.80
    vlan_link: eth0
    vlan_id: 80
    subnets:
      - type: static
        address: 10.10.80.20/24
        gateway: 10.10.80.1
  - type: nameserver
    address:
      - 10.10.80.1
    search:
      - home.m-koetter.de
```

If custom NTP servers are needed, this can be done in `user-data` for example:

```
ntp:
  enabled: True
  servers: ["10.10.80.1"]

runcmd:
  - "systemctl restart ntp"
```

Not sure whether these tweaks could be done in `os-rootfs` - only the Pi image supports `cloud-init`, right?

**Comment by RenEvo** [comment]
About to jump in and start doing a lot more cloud-init, this looks like a good fix!

**Comment by pret79** [comment]
Any update when it is planned to add PR #247 to Hypriot? I'd also appreciate having support for VLAN interfaces via the _network-config_ file.

**Comment by 616b2f** [comment]
@michaelkoetter wow thank you, spend some time with the same issues, that helped me a lot! It would be also a nice PR to https://github.com/DieterReuter/image-builder-rpi64 I used this for my rpi k8s cluster and had the same issues.

**Comment by StefanScherer** [comment]
Ok, YOLO, let's merge this.

**Comment by Legion2** [comment]
When will be the next version of hypriot be released so we can test this?

PS: The circleci pipeline is currently failing, so I can't download the image from there

**Comment by StefanScherer** [comment]
@Legion2 I kicked off a build today, Circle is green again. I‘ll draft a release later this week.

**Comment by rbartoli** [comment]
@StefanScherer It seems that since merging this PR the configuration of a static ip address on eth0 through `user-data` file does not work anymore (as I've reported in #359).

I'm not familiar with network-config, so by looking at the code of the PR I wondered:
- As I still see `user-data` being used, and `network-config` added after, is `network-config` currently forcing `dhcp` on `eth0` through this line: https://github.com/hypriot/image-builder-rpi/blob/4b5348a95ccee1e3c924d9c5c371bc8e075070f4/builder/files/boot/network-config#L6?
- Does this PR introduce a new way of setting a static ip through a `network-config`? I may have missed some documentation updates but I did search.

**Comment by RenEvo** [comment]
@rbartoli https://cloudinit.readthedocs.io/en/latest/topics/network-config-format-v2.html

```
network:
  version: 2
  ethernets:
    eth0:
      dhcp4: false
      dhcp6: false
      addresses: [10.0.0.40/24]
      gateway4: 10.0.0.1
      nameservers:
        search: [foo.local, bar.local]
        addresses: [8.8.8.8]
```

This should give kind of what you expect.

**Comment by LS80** [comment]
This PR also breaks adding dns servers via `dhcpcd` with
```
static domain_name_servers=127.0.0.1 1.1.1.1 1.0.0.1
```
in `/etc/dhcpcd.conf`.

I have tried adding nameservers with the `network-config` but that is not updating `/etc/resolv.conf`.
